src/main/host/syscall/mod.rs

```diff
@@ -72,3 +74,115 @@ pub fn get_descriptor(
 
     Ok(desc)
 }
+
+pub enum ResolvedPluginPtr<T> {
+    /// Was `NULL` inside the plugin.
+    Null,
+    /// Length was 0.
+    ZeroLen,
+    /// Length was too small to represent the generic type `T`.
+    LenTooSmall,
+    /// A local aligned non-null address for the plugin data.
+    Address(*const T),
+}
+
+pub enum MutResolvedPluginPtr<T> {
+    /// Was `NULL` inside the plugin.
+    Null,
+    /// Length was 0.
+    ZeroLen,
+    /// Length was too small to represent the generic type `T`.
+    LenTooSmall,
+    /// A local aligned non-null address for the plugin data.
+    Address(*mut T),
+}
+
+pub fn get_readable_ptr<T>(
```

Comment:
It'd be nice if we could encapsulate the raw pointers into a smaller piece of the code. The MemoryManager has some APIs like this (`get_ref` etc) but those don't fall back to the Thread methods on a miss.

Would it make sense to introduce a Process object here on the Rust side that wraps a Process*, and provides more idiomatic Rust apis? (Or maybe a Process *trait*, with an implementation that wraps a C process, as we have with Thread)

The return types could use `Result` with the appropriate reference as the `Ok` type, and a single custom error type enum replacing the two you have above.

This will potentially force us to deal with some ownership and mutability issues, but we probably *want* that rather than giving up safety by working with raw pointers. (If we really wanted to dodge those, I think maybe these APIs could return references with 'static lifetime instead of the lifetime of the Process object, but that'd again be giving up some safety)

Reply:
> It'd be nice if we could encapsulate the raw pointers into a smaller piece of the code. The MemoryManager has some APIs like this (get_ref etc) but those don't fall back to the Thread methods on a miss.

There are three main issues:
1. Like you mentioned, we need to access a path that has the fallback option, so we can't access the memory manager directly.
2. The MemoryManager's methods like get_ref() and get_slice() don't work well for types like `sockaddr` which later has to be re-interpreted as types of possibly larger sizes. We could get a u8 slice, and then convert the bytes to the various sockaddr types, but this doesn't really seem any safer than just using a pointer. Is there a nicer way of dealing with this?
3. You cannot borrow more than one reference at a time from the memory manager. You have a comment `TODO: Consider tracking borrowed memory ranges so that we can safely allow mutable borrowed references, with a run-time check [...]` so we would need to implement that.

I think if we want safety here, we need to restructure how we manage plugin memory. We need to remove the implicit `thread_flushPtrs()` so that we can give the references proper lifetimes. We need a manager for the MemoryManager and the fallback path to make sure that they don't return aliased pointers and that they're aligned. Or we'd need to move the fallback path out of `process_getWriteablePtr()` and into MemoryManager, and implement this runtime bookkeeping in MemoryManager instead. This should also be compatible with the C code.

> Would it make sense to introduce a Process object here on the Rust side that wraps a Process*, and provides more idiomatic Rust apis? (Or maybe a Process trait, with an implementation that wraps a C process, as we have with Thread)

We'd want to avoid ever creating two Process objects with the same pointer. I'm not sure how we can guarantee this. How about if the C Process object has a `bool rustWrapperExists` and the rust Process object sets this when created, and removes it when dropped?

> The return types could use Result with the appropriate reference as the Ok type, and a single custom error type enum replacing the two you have above.
>
> This will potentially force us to deal with some ownership and mutability issues, but we probably want that rather than giving up safety by working with raw pointers. (If we really wanted to dodge those, I think maybe these APIs could return references with 'static lifetime instead of the lifetime of the Process object, but that'd again be giving up some safety)

We could do this, but we'd need to figure out the memory management issue first. I don't think we should return a reference with a `'static` lifetime as that seems more dangerous to me than returning a pointer.

Reply:
> Like you mentioned, we need to access a path that has the fallback option, so we can't access the memory manager directly.

Yup, I think we'll want to expose the Process APIs in rust and wrap them so they return references.

> The MemoryManager's methods like get_ref() and get_slice() don't work well for types like sockaddr which later has to be re-interpreted as types of possibly larger sizes. We could get a u8 slice, and then convert the bytes to the various sockaddr types, but this doesn't really seem any safer than just using a pointer. Is there a nicer way of dealing with this?

Good point. Still, I think it's better to return references that can just be used as references in the common case, and only in some exceptional cases have the handling code have to delve back into casting/transmuting. Alternatively we could still provide APIs that return pointers, so the syscall handlers that do need pointers could just use those, but I think that'd be weakening any exclusivity guarantees etc.

> You cannot borrow more than one reference at a time from the memory manager. You have a comment TODO: Consider tracking borrowed memory ranges so that we can safely allow mutable borrowed references, with a run-time check [...] so we would need to implement that.

Maybe. I have been wondering if we can get away without it. A syscall handler could just "check out" one reference at a time and return it before getting the next. For pointers to primitives or smallish structs it could just copy locally before releasing the reference if needed.

> We could do this, but we'd need to figure out the memory management issue first. I don't think we should return a reference with a 'static lifetime as that seems more dangerous to me than returning a pointer.

Fair enough.

I'm on the fence whether it makes sense to try to add these APIs now. On one hand it's nontrivial and will block your work in progress, on the other hand doing it now would save work in the long run vs. implementing more handlers that operate on raw pointers and then migrating them to reference-based APIs later.